CopyObject api and gateway using copy-table tracking

nopcoder opened the pull request:
Close #4477

itaiad200 commented on pkg/catalog/catalog.go:
```diff
@@ -1923,6 +1930,28 @@ func (c *Catalog) PrepareGCUncommitted(ctx context.Context, repositoryID string,
+		return "", err
+	}
+	repoPartition := graveler.RepoPartition(repo)
+	id := xid.New().String()
+	key := kvTrackPrefix + id
```
Why can't the physical address (prefixed by the repo) be the key?

nopcoder replied:
wanted to sort order to be ascending order - our physical addresses ordered by desc.

itaiad200 commented on pkg/catalog/catalog.go:
```diff
@@ -1940,29 +1969,116 @@ func (c *Catalog) VerifyLinkAddress(ctx context.Context, repository, token strin
+}
+
+func (c *Catalog) deleteRepoTrackedPhysicalAddresses(ctx context.Context, repo *graveler.RepositoryRecord) error {
+	// collect and delete old keys - work in 'kvTrackDeleteBatchSize' and interval of 5 min between if batch is bigger
```
You can skip the optimizations for kv calls at the moment, we need to handle this as part of the bigger picture. This is not the only place where this pose a problem.

nopcoder replied:
Suggest to remove the sleep or make the delete unbound to any size?

itaiad200 commented on pkg/catalog/catalog.go:
```diff
@@ -1940,29 +1969,116 @@ func (c *Catalog) VerifyLinkAddress(ctx context.Context, repository, token strin
+func (c *Catalog) deleteRepoTrackedPhysicalAddressHelper(ctx context.Context, repo *graveler.RepositoryRecord) (bool, error) {
+	repoPartition := graveler.RepoPartition(repo)
+	msgType := (&Entry{}).ProtoReflect().Type()
+	it, err := kv.NewPrimaryIterator(ctx, c.kvStore.Store, msgType, repoPartition, []byte(kvTrackPrefix), kv.IteratorOptionsFrom([]byte{}))
```
You're always starting from the top. 
Every 2 seconds you will be getting the same results which should not be deleted, it's a waste in kv calls

itaiad200 replied:
Oh, is that the reason for the sorted keys?

nopcoder replied:
yes, I want to delete old 1k keys and repeat the action without overload the system with delete requests.

itaiad200 commented on pkg/api/controller.go:
```diff
@@ -2334,85 +2338,156 @@ func (c *Controller) StageObject(w http.ResponseWriter, r *http.Request, body St
+
+	// create entry (copy) - try only once
+	dstEntry := *srcEntry
+	dstEntry.CreationDate = time.Now().UTC()
```
Isn't the `.UTC()` redundant?

nopcoder replied:
Removed as CreateEntry handles that - thanks.

itaiad200 commented on pkg/api/controller.go:
```diff
@@ -2334,85 +2338,156 @@ func (c *Controller) StageObject(w http.ResponseWriter, r *http.Request, body St
+	dstEntry := *srcEntry
+	dstEntry.CreationDate = time.Now().UTC()
+	dstEntry.Path = destPath
+	err = c.Catalog.CreateEntry(ctx, repository, branch, dstEntry, graveler.WithMaxTries(1))
```
Can you document the `1` try limitation?

itaiad200 commented on pkg/api/controller.go:
```diff
@@ -2334,85 +2338,156 @@ func (c *Controller) StageObject(w http.ResponseWriter, r *http.Request, body St
+
+	// copy data to a new physical address
+	dstEntry := *srcEntry
+	dstEntry.CreationDate = time.Now().UTC()
```
Same here

itaiad200 commented on pkg/api/controller.go:
```diff
@@ -2334,85 +2338,156 @@ func (c *Controller) StageObject(w http.ResponseWriter, r *http.Request, body St
+	var err error
+	// fetch src entry if needed - optimization in case we already have the entry
+	if srcEntry == nil {
+		srcEntry, err = c.Catalog.GetEntry(ctx, repo.Name, branch, srcPath, catalog.GetEntryParams{ReturnExpired: true})
```
Why do we return expired?

nopcoder replied:
This is the default for get object - but looking into it, no one is using this field so I'll handle it in a different PR.

nopcoder replied:
Handled in this PR

itaiad200 commented on pkg/api/controller_test.go:
```diff
@@ -3860,64 +3860,157 @@ func TestController_CopyObjectHandler(t *testing.T) {
+		require.Nil(t, deep.Equal(statResp.JSON200, copyStat))
+	})
+
+	t.Run("full_diff_branch", func(t *testing.T) {
```
diff has many meanings
```suggestion
	t.Run("full_different_branch", func(t *testing.T) {
```

itaiad200 commented on pkg/gateway/operations/putobject.go:
```diff
@@ -136,38 +96,115 @@ func getPathFromSource(copySource string) (path.ResolvedAbsolutePath, error) {
 	}
+	ctx := req.Context()
+
+	// TODO(barak): verify that o.Reference must be a branch
```
Delete or fix

itaiad200 commented on pkg/gateway/operations/putobject.go:
```diff
@@ -136,38 +96,115 @@ func getPathFromSource(copySource string) (path.ResolvedAbsolutePath, error) {
 	}, http.StatusOK)
 }
 
+func copyObjectShallow(ctx context.Context, c catalog.Interface, repository string, branch string, srcEntry *catalog.DBEntry, destPath string) (*catalog.DBEntry, error) {
```
Can't we reuse `copyObjectShallow` & `copyObjectFull`?
Not blocking but I think it's worth a scoped little effort if possible

itaiad200 commented on pkg/gateway/operations/base.go:
```diff
@@ -39,7 +39,7 @@ const (
 )
 
-type ActionIncr func(action, repository, ref, userID string)
+type ActionIncr func(action, userID, repository, ref string)
```
Did you change all the usages? I missed it

nopcoder replied:
This is a fix - all the calls use this order of arguments.

itaiad200 reviewed (changes requested):
Thanks for this awesome almost-last-piece of the ugc#2 puzzle! Great tests too!
Most comments around questions and clarifications, nothing too serious but they piled up so requesting changes.

nopcoder commented on pkg/graveler/graveler.go:
```diff
@@ -1522,17 +1553,16 @@ func (g *Graveler) DeleteBatch(ctx context.Context, repository *RepositoryRecord
-				if err != nil {
-					m = multierror.Append(m, &DeleteError{Key: key, Err: err})
-				}
+	err = g.safeBranchWrite(ctx, g.log(ctx).WithField("operation", "delete_keys"), repository, branchID, 0, func(branch *Branch) error {
```
I'll pass options struct so it will be clear to the reader

nopcoder commented:
@itaiad200 addressed all the comments, except the one related to the delete of expired tracked physical addresses.
waiting for feedback.
Note that this pr is based on @N-o-Z 's work - thanks for the kind words.

itaiad200 reviewed (approved):
Thanks for making the changes!
